### PR TITLE
Ensemblers CRUD API

### PR DESCRIPTION
**Pull request by romanwozniak**
#### TL;DR
This PR adds CRU(D) API for Ensemblers, together with some refactoring around Controllers layer of the turing-api. Also, it contains WIP [openapi 3.0 spec](https://github.com/gojek/turing/compare/main...romanwozniak:log_ensembler_api?expand=1#diff-88f51503d6a9fc66f6f7f37fd8cefe47ff2a1fe394c02f606889d9ba4b7620e7) for the batch ensembling jobs.

#### Details
Turing is progressing towards adding support of the batch ensembling jobs. The ensembling logic for batch use-cases is expected to be implemented by users extending base Python Ensembler (to be added into Turing SDK) and saving their `PyFunc` Ensemblers in Turing backend. Then, users would be able to submit an ensembling job by providing a configuration of the source/sink of the data and information about which ensembler should be used for doing the ensembling. 

With that in mind, we are introducing `Ensemblers` as another top-level abstraction in Turing (in addition to existing `Routers`). Currently, the only supported type of stored Ensemblers is `pyfunc` (this will later be extended to support `docker` ensemblers as well).

#### Notable Changes:
- Existing DB table `ensemblers` is renamed to `ensembler_configs`
- The signature of Controller methods is changed from
    ```go
    Method(*http.Request, map[string]string, interface{})
    ```
     to
    ```go
    type RequestVars map[string][]string
    
    Method(*http.Request, RequestVars, interface{}) 
    ```
    to make request variables (both extracted from the req's path and query params) to be the same type as `*http.Request.URL.Query()`
- Definitions of API routes have been moved from `api/handler.go` into separate controllers (e.g. `api/routers_api.go`, `api/ensemblers_api.go`)
- Fixed `RoutersService.Delete(...)` method that previously wasn't properly cleaning out corresponding Enricher/Ensembler configurations, belonging to this router versions. (+fixed tests that didn't capture that issue earlier) 
- Added `models.ID` in attempt to standardize type of IDs that are used in turing-api (not 100% sure if I like it, so maybe this will be changed later)
- Added offset based pagination into `EnsemblersService.List`

### CODE REVIEW

File: api/swagger-batch.yaml [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+              description: "Total number of results matching the query criteria"
+            page:
+              type: "integer"
+              format: "int32"
+              description: "Number of the current page"
+            pages:
+              type: "integer"
+              format: "int32"
+              description: "Total number of pages"
+
```
Comment (krithika369) [inline api/swagger-batch.yaml:310]: Since a number of these data structures are meant to match the [Proto](https://github.com/gojek/turing/blob/main/engines/batch-ensembler/api/proto/v1/batch_ensembling_job.proto), I'm wondering if it would make sense to auto-generate them using a library like [this](https://github.com/google/gnostic) and reference them by splitting the OpenAPI specs.

**Review by krithika369 — COMMENTED** [review]
Added some comments on the OpenAPI specs. I will review rest of the MR later today.

File: api/turing/models/pyfunc_ensembler.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+	SetProjectID(id ID)
+	Patch(other EnsemblerLike) error
+}
+
+func EnsemblerTable(ensembler EnsemblerLike) func(tx *gorm.DB) *gorm.DB {
+	return func(tx *gorm.DB) *gorm.DB {
+		switch ensembler.GetType() {
+		case EnsemblerTypePyFunc:
+			return tx.Table("pyfunc_ensemblers")
+		default:
```
Comment (krithika369) [inline api/turing/models/pyfunc_ensembler.go:27]: Is this expected? Since there is no ensembler config related info in the `ensemblers` table, I imagine that table is just used to store the common fields. Should the default case raise an error?

**Review by krithika369 — COMMENTED** [review]
Hi @romanwozniak, I'm done with my review. Thanks for the extensive and thoughtful refactoring. Added some more questions.

File: api/swagger-batch.yaml [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+          name: page
+          schema:
+            type: integer
+            default: 1
+        - in: query
+          name: page_size
+          schema:
+            type: integer
+            default: 10
+      responses:
```
[the remark this answers is not in the document]
Comment (romanwozniak) [reply api/swagger-batch.yaml:28]: The error codes are likely the same (or largely the same) for all the endpoints, hence they aren't that useful. This spec is WIP and will likely be merged with the main [swagger spec](https://github.com/gojek/turing/blob/main/api/swagger.yaml) later, the error codes can probably be documented then as well

File: api/db-migrations/000008_pyfunc_ensemblers.down.sql [hunk 4]
```diff
@@ -0,0 +1,5 @@
+ALTER TABLE ensembler_configs RENAME TO ensemblers;
+
+DROP TABLE IF EXISTS pyfunc_ensemblers;
+
+DROP TABLE IF EXISTS ensemblers;
```
[the remark this answers is not in the document]
Comment (romanwozniak) [reply api/db-migrations/000008_pyfunc_ensemblers.down.sql:5]: Right, thanks for catching 🙌

File: api/swagger-batch.yaml [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+        "201":
+          description: A JSON representation of a successfully saved ensembler
+          content:
+            application/json:
+              schema:
+                oneOf:
+                  - $ref: "#/components/schemas/PyFuncEnsembler"
+
+  "/projects/{project_id}/ensemblers/{ensembler_id}":
```
[the remark this answers is not in the document]
Comment (romanwozniak) [reply api/swagger-batch.yaml:68]: Yeah, it is intentional. I wasn't sure what subsequences will it have if a user removes the ensembler, that was used for submitting an ensembling job (the API for that is WIP by @ashwinath), so I skipped the deletion for now

File: api/swagger-batch.yaml [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+            application/json:
+              schema:
+                allOf:
+                  - $ref: "#/components/schemas/PaginatedResults"
+                  - type: object
+                    properties:
+                      results:
+                        type: array
+                        items:
+                          $ref: "#/components/schemas/GenericEnsembler"
```
[the remark this answers is not in the document]
Comment (romanwozniak) [reply api/swagger-batch.yaml:41]: 1. I don't think that this is required from the user-flow perspective (i.e. there would probably a view on UI for listing all ensemblers, so it would only render common properties of them).
2. It's not easy to implement this in gorm (or more accurately, generally in PQ), because you can't select from Parent AND child tables in a single query (you can't do a UNION of tables with a different column structure as well)

File: api/swagger-batch.yaml [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+          type: "string"
+        annotations:
+          type: "object"
+          additionalProperties:
+            type: "string"
+
+    EnsemblingJobSpec:
+      description: Ensembling job Specification
+      properties:
+        source:
```
[the remark this answers is not in the document]
Comment (romanwozniak) [reply api/swagger-batch.yaml:197]: Simply because I haven't added it :) The `jobs` spec is out of scope of this PR, but I just started adding some draft for sharing the context with @ashwinath, this can be extended as the part of his PR

File: api/swagger-batch.yaml (hunk 1 above)
Comment (romanwozniak) [reply api/swagger-batch.yaml:310]: Interesting suggestion! :) 
When I defined the original [proto](https://github.com/gojek/turing/blob/main/engines/batch-ensembler/api/proto/v1/batch_ensembling_job.proto), I spent some time thinking about how to reduce this repetitive effort of maintaining both openapi spec and proto spec, but haven't decided how exactly to approach it. In my opinion, there is no much benefits of maintaining the proto spec, it can probably be replaced with openapi spec alone. But since ensembling job is in scope of `engines/batch-ensembler`, and openapi spec belongs to `api`, it made it hard to merge them easily. 
Maybe in the future I'll just move openapi specs into the root level of the project, remove the proto and use [openapi-generator](https://github.com/OpenAPITools/openapi-generator) to generate python/golang code of these data structures.

File: api/turing/api/alerts_api.go [hunk 8]
```diff
@@ -10,12 +10,12 @@ import (
...
 
 type AlertsController struct {
-	*baseController
+	*BaseController
 }
 
 var ErrAlertDisabled = errors.New("alert is disabled in turing-api")
 
-func (c *AlertsController) CreateAlert(r *http.Request, vars map[string]string, body interface{}) *Response {
+func (c AlertsController) CreateAlert(r *http.Request, vars RequestVars, body interface{}) *Response {
```
[the remark this answers is not in the document]
Comment (romanwozniak) [reply api/turing/api/alerts_api.go:18]: 1. It's much easier: I did it for the consistency with other controllers (for example, [outers_api.go](https://github.com/gojek/turing/blob/main/api/turing/api/routers_api.go)). Generally, I can't see any reasons of why it should be a pointer receiver (i.e, mutability is not required, performance is not a concern, etc.)
2. Because BaseController is also defined with pointer receivers (I missed that) and initially I added `Routes() []Route` [method](https://github.com/romanwozniak/turing/blob/a05dc201c839a036e3c2f05c24939fe22d46d7d7/api/turing/api/base_controller.go#L12) into it to implement `Controller` interface and stub methods of that interface for all the controllers that have BaseController embedded. Since `BaseController` is defined with pointer receivers, it should be embedded by pointer as well. 
I will just update `BaseController` to use value receivers.

File: api/turing/models/ensembler.go [hunk 9]
```diff
@@ -14,11 +14,16 @@ type Ensembler struct {
 	DockerConfig   *EnsemblerDockerConfig   `json:"docker_config"`   // Ensembler config when Type is "docker"
 }
 
+func (*Ensembler) TableName() string {
```
[the remark this answers is not in the document]
Comment (romanwozniak) [reply api/turing/models/ensembler.go:17]: I have added a doc comment answering your question. Here it is:
> TableName returns the name of a table, where GORM should store/retrieve 
entities of this type. By default GORM uses the table name, that is 
a plural form of the type's name (i.e 'Ensembler' -> `ensemblers`),
and by implementing `TableName` method it is possible to override it.

File: api/turing/models/pyfunc_ensembler.go [hunk 10]
```diff
@@ -0,0 +1,113 @@
+package models
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/jinzhu/gorm"
+)
+
+type EnsemblerLike interface {
```
[the remark this answers is not in the document]
Comment (romanwozniak) [reply api/turing/models/pyfunc_ensembler.go:10]: Becoz `Ensembler` is taken by the [Ensembler](https://github.com/gojek/turing/blob/092839fdaf8622ebb7770f93763ac27d0b09f4d2/api/turing/models/ensembler.go#L10). I will rename it to `Ensembler` once I come up with a better name for the Ensembler configurations, that are used in router deployments (`ensembler_configs` is just a WIP version)

File: api/turing/models/pyfunc_ensembler.go (hunk 2 above)
Comment (romanwozniak) [reply api/turing/models/pyfunc_ensembler.go:27]: From the service perspective (and generally from the inherited db tables perspective) – this is expected. From the application perspective this doesn't really make sense (because generic ensemblers aren't useable), and that's why we have [validation](https://github.com/romanwozniak/turing/blob/a05dc201c839a036e3c2f05c24939fe22d46d7d7/api/turing/api/ensemblers_api.go#L174) that checks the type of ensembler and rejects ensembler with unsupported types

File: api/turing/service/router_service.go [hunk 11]
```diff
@@ -112,21 +116,28 @@ func (service *routersService) Delete(router *models.Router) error {
...
 	tx := service.db.Begin()
-	var routerVersions []*models.RouterVersion
 
 	// remove associations
 	router.ClearCurrRouterVersion()
 	tx.Save(router)
 
-	service.db.Model(router).Related(&routerVersions)
+	var routerVersions []*models.RouterVersion
+	if err := service.db.
```
[the remark this answers is not in the document]
Comment (romanwozniak) [reply api/turing/service/router_service.go:125]: 1. Side note: we are on gorm v1, so the correct version of docs is here: https://v1.gorm.io/docs/preload.html
This is the first time I hear about `when it is accessed` and I'm not even sure if it's possible to easily implement such pattern in golang. (I've seen it before in Hibernate with Java, but that's a different story). So it's not a performance optimization, but rather an instruction for gorm to either load the assosiated entry(ies) or not. 
With that said, if preloading doesn't happen, then corresponding `routerVersion.Enricher` and `routerVersion.Ensembler` are always `nil`, hence – not removed from the database. So your assumption leads you by the wrong path.
> And if so, is it not enough to always delete the router version only after the enricher/ensembler is deleted? 
This is wrong too. `router_versions` table references `enrichers.id`/`ensemblers.id`, so you can't delete Enricher/Ensembler first, because in this case the record in `router_versions` table will reference non-existing records in `enrichers`/`ensemblers` tables, which is prohibited. 

But now, I realized that we don't really need to `Preload` `Enricher`/`Ensembler` and we can just use `EnsemblerId`/`EnricherId` and delete corresponding records by their primary keys. So, thanks for bringing my attention to this :)

2. Hooks in general are harder to navigate and understand, so I'd prefer the (current) explicit deletion method.

**Comment by romanwozniak** [comment]
@krithika369 thank you for review! I've addressed/answered all of your comments 🙌

File: api/swagger-batch.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (ashwinath) [reply api/swagger-batch.yaml:41]: I'm not sure if we should standardise the types of ensemblers returned here.

Right now the problem seems to be the GET and POST method seems to support different ensemblers. GET giving generic but POST only allowing pyfunc ensemblers. I think it would be easier in the future to support a field called `type` and every endpoint returning a GenericEnsembler type with a field called `metadata` that contains the specific metadata inside.

I might not be getting the pattern here but it seems difficult to extend this API to support other types of ensemblers in the future. I am guessing thats the reason why we are doing generic + subtype ensemblers here.

[the remark this answers is not in the document]
Comment (ashwinath) [reply api/swagger-batch.yaml:41]: I'm not too familar with the gorm library but i think it can be easily done with some if else statements to select from which tables depending on the type of ensembler.

File: api/swagger-batch.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (ashwinath) [reply api/swagger-batch.yaml:68]: The complication would arise when the job has been submitted and running/queued up since it's running in the background. Extra check logic needs to be done to ensure that the job should run if the ensembler is missing. While this can be done, I would suppose that it is a matter of whether it is worth the time implementing such a feature.

File: api/swagger-batch.yaml [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+      allOf:
+        - $ref: "#/components/schemas/GenericEnsembler"
+        - type: "object"
+          properties:
+            mlflow_url:
+              type: "string"
+              format: "uri"
+            mlflow_experiment_id:
+              $ref: "#/components/schemas/Id"
+            mlflow_run_id:
```
Comment (ashwinath) [inline api/swagger-batch.yaml:171]: Curious, what are the mlflow url/experiement_id and run_id used for this ensembling job? My understanding is that we only require the artifact_uri to build the ensembler and run it.

File: api/swagger-batch.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (ashwinath) [reply api/swagger-batch.yaml:197]: Yeah I'll handle this in my PR. There's going to be a lot more than that.

File: api/swagger-batch.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (krithika369) [reply api/swagger-batch.yaml:68]: For the routers / versions, we have additional checks to ensure there are no other references to an entity we are trying to delete (or, in some cases, all associated entities are deleted together):
* Can't delete currently deployed router version
*  Can't delete a router version if that's the only version under a router (the router must be deleted directly)
* ...
So, to @ashwinath's point on "Extra check logic needs to be done to ensure that the job should run if the ensembler is missing." - perhaps the delete can be disalowed if there is a job in progress referencing it.

Anyway, sounds like Delete has not been added yet for these considerations. Thanks for the clarification.

File: api/swagger-batch.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (krithika369) [reply api/swagger-batch.yaml:41]: > I might not be getting the pattern here but it seems difficult to extend this API to support other types of ensemblers in the future.

@ashwinath I think the idea is to keep the `oneOf` updated with new types of ensemblers (but Roman can CMIIW). I echo your concern on the inconsistency between the endpoints.

@romanwozniak 
1. Would the UI need to show all types of ensemblers together? Or can we perhaps add a required `type` filter to the query?
2. I thought it could be multiple queries - one for each concrete type of ensemblers that is combined on the Golang side. But then the ordering, pagination, etc. are an issue. 🤔 I think if we really want to, it can still be done (query the generic table first and issue subsequent queries to the individual tables and and combine programmatically). But it is more complex.

Now thinking about it more, why are we not following the single table approach we've been using for the real-time Ensemblers (or the ExperimentEngine model)?
```
type: ...
standard_config: ...
docker_config: ...
```

File: api/turing/api/ensemblers_api.go [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+
+	return Ok(results)
+}
+
+func (c EnsemblersController) GetEnsembler(
+	_ *http.Request,
+	vars RequestVars,
+	_ interface{},
+) *Response {
+	params := struct {
```
Comment (ashwinath) [inline api/turing/api/ensemblers_api.go:42]: Why not declare it as a named struct instead of defining it every HTTP call?

File: api/turing/service/router_service.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (krithika369) [reply api/turing/service/router_service.go:125]: Okay, my assumption that the support for eager loading in Go Gorm should mean there is an implicit lazy loading comes from Django's ORM. I tried looking up docs for Go Gorm and verifying it locally - you are right that there is no "when it is accessed" and rather, it must at least be done with [Related](https://pkg.go.dev/github.com/jinzhu/gorm#DB.Related) at query time.

Thanks for the clarification!

File: api/turing/service/ensembler_service.go [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+	var count int
+	done := make(chan bool, 1)
+
+	query := service.db
+	if options.ProjectID != nil {
+		query = query.Where("project_id = ?", options.ProjectID)
+	}
+
+	go func() {
+		query.Model(&results).Count(&count)
```
Comment (ashwinath) [inline api/turing/service/ensembler_service.go:73]: I'm curious to know how this will turn out. From my understanding, value of results could be in a indeterminate state when line 79 executes as well. I'm not sure if this will have an effect when the value of results changes when line 79 executes.

File: api/swagger-batch.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (romanwozniak) [reply api/swagger-batch.yaml:68]: The other way of dealing with it would be to treat records from `ensemblers` table some sort of templates, and create a snapshot config from these templates at the deployment time (either router deployment / job submission). So in this case, the actual record from the `ensemblers` table can be safely deleted at any time, because the job will use a copy of it's data.
But I'll keep it as is for now and get back to this question once the jobs API is ready

File: api/swagger-batch.yaml (hunk 1 above)
Comment (romanwozniak) [reply api/swagger-batch.yaml:310]: Just to make it clear, I don't plan to work or on this idea in this PR :)

File: api/swagger-batch.yaml (hunk 12 above)
Comment (romanwozniak) [reply api/swagger-batch.yaml:171]: Because Python ensemblers are in fact a specific implementation of MLFlow Python models and are packaged (pickled) and uploaded to the artefact store (GCS) by mlflow client. So this makes it consistent with merlin pyfunc models as well. 
You wouldn't require these fields for the job submission flow

File: api/turing/service/ensembler_service.go (hunk 14 above)
Comment (romanwozniak) [reply api/turing/service/ensembler_service.go:73]: Oh, this is not how gorm works. `&results` here are only used to tell gorm the type of entities (and subsequently, the name of table where it should query the data from). This is effectively the same as `query.Table("ensemblers").Count(&count)` or `query.Model(models.GenericEnsembler{}).Count(&count)`

File: api/turing/api/ensemblers_api.go (hunk 13 above)
Comment (romanwozniak) [reply api/turing/api/ensemblers_api.go:42]: I declared it as a named struct first, but that makes the validation error even less relevant (something like `Key: 'NamedStruct.EnsemblerID' Error:Field validation for 'NamedStruct.EnsemblerID' failed on the 'required' tag`). But I guess it doesn't matter that much, because errors returned by API are not translated into user-readable messages anyway.

File: api/swagger-batch.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (romanwozniak) [reply api/swagger-batch.yaml:41]: > Right now the problem seems to be the GET and POST method seems to support different ensemblers

@ashwinath I'm probably not getting the problem you are describing here. GET method (`projects/:project_id/ensemblers/:ensembler_id`) returns exactly the same data as was written into the database from POST `projects/:project_id/ensemblers`. GET `projects/:project_id/ensemblers` returns only generic data of each ensembler, but that is fine too, because the flow would be `List all ensemblers` -> `Click on a specific ensembler` -> `See the details of that ensembler` and List view would only need generic properties of each ensembler.

[the remark this answers is not in the document]
Comment (romanwozniak) [reply api/swagger-batch.yaml:41]: > Now thinking about it more, why are we not following the single table approach we've been using for the real-time Ensemblers (or the ExperimentEngine model)?

@krithika369 I considered it, but preferred the table inheritance approach because it maps more naturally to the types composition in golang (i.e., I like the idea of having a `PyFuncEnsembler` struct with all the data available in it, instead of having a struct `Ensembler` and then look for `pyfunc_config` in it to check the data). It's probably possible to achieve the same with a single table approach and custom DB Write/Read methods, but that seems more complicated

File: api/turing/api/ensemblers_api.go (hunk 13 above)
Comment (romanwozniak) [reply api/turing/api/ensemblers_api.go]: Done that here: https://github.com/gojek/turing/pull/55/commits/0f33f0e269f768fe9d7a008345c24d6e0385e0d2

**Comment by romanwozniak** [comment]
@ashwinath @krithika369 is there anything else I can answer regarding this PR? If no, then don't forget to add your approval :)

**Review by krithika369 — APPROVED** [review]
LGTM 🎉
